elasticai/creator/ir/graph_iterators.py

```diff
@@ -0,0 +1,35 @@
+from collections.abc import Callable, Iterable, Iterator
+from typing import Hashable, TypeAlias, TypeVar
+
+HashableT = TypeVar("HashableT", bound=Hashable)
+
+NodeNeighbourFn: TypeAlias = Callable[[HashableT], Iterable[HashableT]]
+
+
+def dfs_pre_order(successors: NodeNeighbourFn, start: HashableT) -> Iterator[HashableT]:
+    visited: set[HashableT] = set()
+
+    def visit(nodes: tuple[HashableT, ...]):
+        for n in nodes:
+            if n not in visited:
+                yield n
+                visited.add(n)
+                yield from visit(tuple(successors(n)))
+
+    yield from visit((start,))
```

Comment:
Pleb question, what is happening here ? :D

Reply:
Hmm... maybe the function shouldn't be called `visit`. The algorithm only needs to retrieve successors of the current node, so we pass a function and the start node to it.

Lemme try to explain, maybe that leads to better naming, pseudo code on:

```
function(n)

  for each successor of n:
    if successor has not been visited:
      register successor as visited
      pass successor to caller
      pass all values from function(successor) to caller

pass all values from function("start") to caller
      
```

Reply:
So yeah, could rewrite it to say ask for successors in the head of the for loop... Not sure if any that helps with your question.

Another info:
  swapping lines 15 and 17 would make it post-order

Reply:
No I think recursive is fine, the naming too. I was just a little confused with python syntax but I got it now x) That's why I warned that it's a pleb question